frontend/src/common/queries/filter.ts

```diff
@@ -188,17 +206,27 @@ export function useFetchDatasetRows(): FetchCategoriesRows<DatasetRow> {
     isLoading: collectionsLoading,
   } = useFetchCollections();
 
-  // Build dataset rows once datasets and collections responses have resolved.
+  // Build dataset rows once datasets and collections responses have resolved. Update loading and error fetch states.
   useEffect(() => {
+    setIsLoading(datasetsLoading || collectionsLoading);
+    setIsError(datasetsError || collectionsError);
+
     if (!datasets || !collectionsById) {
       return;
     }
     setDatasetRows(buildDatasetRows(collectionsById, datasets));
-  }, [datasets, collectionsById]);
+  }, [
```

Comment:
Hey @tihuan! 

I have updated React Query's `loading` and `error` flags to be state here in an attempt to have atomic updates of state and data where possible. 

Previously, when React Query completed the fetch, I was seeing:

1. `loading` updated from from `true` to `false`.
2. Calling component re-rendered with updated `loading` value.
3. `datasetRows` calculated in this effect.
4. Calling component re-rendered with updated `rows` value.

This flow was causing a brief flash of the "no results" hero in the calling component where `loading` is updated to `false` but the rows haven't yet been computed.

I was hoping to get your opinion on my fix here as I am wondering if there's a cleaner solution than adding React Query flags as state and having such a long dependency list on the hook.

Reply:
Thanks for the PR and describing the issue in great detail, @MillenniumFalconMechanic 🤩 🙏 

It seems like the root cause is `useEffect()` lagging behind by one render cycle, so I have a PR that switches it out to use `useMemo`

https://github.com/chanzuckerberg/single-cell-data-portal/pull/1871

I think the reason is because the states (row state, in this case) that we update inside a `useEffect()` are updated in the next render cycle, that's why when the loading states change, the rows state is one cycle behind

Using `useMemo()` solves the problem, because the calculated result is available immediately and doesn't need to wait till `setState()` to update the value at the next cycle!

Hope that makes sense 😄 !

Reply:
Perfect, makes sense, thanks so much Timmy!